Transform Blocks

illusaen opened the pull request:
## Description
Fixes [#2814](https://github.com/wordpress-mobile/gutenberg-mobile/issues/2814)
Allow users to transform blocks that are:
1. Not explicity disallowed (from block -> to block, i.e. paragraph to group)
2. Explicitly allowed to be transformed from/to (i.e. paragraph)
3. Not unsupported.

*NOTE*: Mobile supports a subset of the web transformations; i.e. we don't support any that the web doesn't, and we don't support all of the web's transformations. Thus, when following the diagram below, the transformations don't match exactly as the web doesn't support all of the transformations on the diagram.

<details>
  <summary>Transformation Category Diagram</summary>

  ![image](https://user-images.githubusercontent.com/13263478/113056654-ac1f5a80-9171-11eb-9207-0b57c7249456.png)

</details>

## How has this been tested?
### Paragraph (allowed)
1. Click paragraph block.
2. Type text in paragraph block.
3. Open block options menu.
4. See that 'Transform block' option exists, and click it.
5. Confirm that 'Heading' exists.
6. Click the 'Heading' button.
7. Confirm that paragraph block has changed to heading block with content preserved.
8. Confirm that there is a banner on top of screen that shows 'Paragraph transformed to Heading'.

### RSS (not allowed)
1. Click RSS block.
3. Open block options menu.
4. See that 'Transform block' option does not exist.

## Screenshots
### New:
https://user-images.githubusercontent.com/13263478/113056224-226f8d00-9171-11eb-9b21-b47e19fe183a.mov


<details>
  <summary>Old GIFs</summary>
  
| iOS | Android |
| - | - |
| <img src="https://user-images.githubusercontent.com/13263478/105678457-cc514580-5eb2-11eb-89c8-272b6ea940a3.gif" width="100%"/> | <img src="https://user-images.githubusercontent.com/13263478/105680752-cdd03d00-5eb5-11eb-86c8-5cf0139c550c.gif" width="78%"/> |
| <img src="https://user-images.githubusercontent.com/13263478/105678465-ceb39f80-5eb2-11eb-9e30-85879e68af15.gif" width="100%"/> | <img src="https://user-images.githubusercontent.com/13263478/105680760-d0cb2d80-5eb5-11eb-95f0-08ad5c75ca81.gif" width="78%"/> |
</details>

## Types of changes
* This code adds transform block functionality to gutenberg-mobile (new feature).

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/native-mobile.md -->

Commit:
WC: WIP: Adding transform block ability with no filters. Also refactored options picker for block settings to have options in one object for better filtering and separated out props to the three different types.

Commit:
Moving getPossibleBLockTransformations to block menu so that we can tell if there are no valid transformation options and if so, not pop up transform option.

Commit:
Adding filter hook to modify transform.isMatch for each registered block depending on if they are allowed to transform.

illusaen commented on packages/block-editor/src/components/block-mobile-toolbar/block-actions-menu.native.js:
```diff
@@ -29,32 +29,42 @@ import { useRef, useState } from '@wordpress/element';
+import BlockTransformationsMenu from '../block-switcher/block-transformations-menu';
 
 const BlockActionsMenu = ( {
-	onDelete,
```
NOTE HERE that I have reordered the props for better clarity. If people hate it I can change it back, but I find it's a lot easier to understand myself when I organize it this way.
1. Props are now organized by either coming from `withDispatch`, `withSelect`, or passed in when created.
2. Props are alphabetically organized within each section.

illusaen commented on packages/block-editor/src/components/block-mobile-toolbar/block-actions-menu.native.js:
```diff
@@ -67,134 +77,138 @@ const BlockActionsMenu = ( {
 		},
 	} = getMoversSetup( isStackedHorizontally, moversOptions );
 
-	const deleteOption = {
```
NOTE: This looks like a huge change, but I'm actually just refactoring to one object instead of having multiple objects, one for each option. This makes it easier to read the code and allows code folding for better readability 🎉 .

illusaen commented on packages/block-editor/src/components/block-mobile-toolbar/block-actions-menu.native.js:
```diff
@@ -67,134 +77,138 @@ const BlockActionsMenu = ( {
-				const serializedBlock = serialize( copyBlock );
+			},
+		},
+		transformButton: {
```
This `transformButton` is the actual change (addition) to the options.

illusaen commented on packages/block-editor/src/components/block-mobile-toolbar/block-actions-menu.native.js:
```diff
@@ -269,16 +292,24 @@ export default compose(
+		);
+
 		return {
+			blockTitle,
```
Note: This is me alphabetizing again.

illusaen commented on packages/block-library/src/index.native.js:
```diff
@@ -182,6 +182,78 @@ addFilter(
+ *   1. Blocks that support transformations
+ *   2. Blocks that aren't in the ineligibleTransformations set
+ *   3. Can only transform to blocks that are supported
+ *     TODO: this should be taken care of by whitelist, no?
```
@chipsnyder This is the specific code relevant to the question I asked in the description: whether or not explicitly checking for unsupported blocks is necessary when there's a whitelist.

chipsnyder replied:
I love that we call out the goals here in the description. 

I agree using the allow list on the `to` and `from` sides seems to meet the goals of `#1` and `#3`.

~I tried various cases to see If I could get a weird scenario, and it all seemed to work nicely.~ (Edit see below) While testing, I noticed that if I add support for `core/group` or `core/columns`, they would be enabled for the heading block. I'll leave that up to @iamthomasbishop if we want to support that or not. but it might be worth adding. `'core/heading': [ 'core/group', 'core/columns' ]` to `ineligibleTransformations` WDYT?

----- 

I had a different question/suggestion related to this. Since we have the `transform.js` files on each block, I wonder if we could utilize a field in there or one added to a possible `transform.native.js` to track the `ineligibleTransformations`. That way, the disallow list is more closely tied to how we handle the allow list. That said, I didn't mess around with that to see what hardships that might introduce. WDYT?

chipsnyder replied:
I did run into one scenario that I think we might want to at least consider. If I add support for `core/group` or `core/columns` then I'll see transform options on Quote (which isn't in the allowed list of `supportedTransformations`). 

I don't know if this has to be fixed now, or even if it's a problem for when we enable a supported transformation, but it might be good to consider a solution and/or leave a note that it's something to be mindful of as we add more transformations we'll need to check all of the blocks it transforms to. WDYT?

illusaen replied:
Hrmm, I think that it is something that needs to be considered. @iamthomasbishop what's the desired behaviour here? If we add in a supported transformation, do we want other (unsupported) blocks to be able to transform to the supported block?

illusaen replied:
@chipsnyder  For the above point: I 100% agree with moving the lists out of this file to jive better with the `transform.js` file; I separated it out for that reason but I honestly wasn't sure where to put it. If I add a `transform.native.js` file, which block would I add it under? Each block has its own `transform.js` file atm so I wasn't sure how to consolidate that.

chipsnyder replied:
I think If I were to describe the goal, I would say: "I don't want paragraph blocks to transform into a group or column blocks" So worded that way, I would probably put an array in the paragraph's `transform.js` or `transform.native.js` file. However, that may not be a great way to make "architectural" decisions 😄. 

As for where it lives in those definitions, I would probably add something at the same level as the to/from definitions.

So in a `transform.native.js` you might have something like
```
const transforms = {
	from: [ ... ],
        to: [ ... ],
	ineligibleTransformations: [ ... ] 
};
```

@gziolo might have an opinion or guidance to add too :)

gziolo replied:
Can you explain what is the goal of this filter? I don't have enough context to answer your question. First, I would have to understand why do you have to apply any change to the list of transformations. Assuming you support all those blocks, why would it be different in the mobile app?

illusaen replied:
@gziolo I think the idea is that it's a lot harder to do certain things on mobile UI, for example, selecting multiple blocks in order to convert to columns/groups.

gziolo replied:
You can still want to convert (in practice wrap) a single block with a group or column. Why would you prevent it?

chipsnyder replied:
Some of the reasons I could see would be to use it as a mechanism to turn off a transformation that for one reason or another is leading to a problematic experience on mobile. Used as a short term safety net. 

@iamthomasbishop you had mentioned wanting to avoid these transformations for v1. I'm not sure if you had any more thoughts there that you would want to add.

gziolo replied:
I can think only about the case when the block you are targeting with transforms is unsupported in the mobile app. Then it doesn't matter how you tackle it in the filter since it's a temp solution. I would only avoid changing the original definition of the block that web uses.

illusaen commented on packages/block-editor/src/components/block-switcher/block-transformations-menu.native.js:
```diff
@@ -0,0 +1,72 @@
+import { withDispatch } from '@wordpress/data';
+
+const BlockTransformationsMenu = ( {
+	// Dispatch
```
See note above about organization of props.

chipsnyder reviewed (commented):
This seems to be working great already! Nice work @illusaen. I had a follow-up question above but I think this is on a good path.

illusaen commented:
Moving white/blacklist to transforms.native.js is taking longer than expected; will come back to this after Simplenote/groundskeeping rotations.

illusaen committed:
Separating out transform logic into transforms.native.js. Adding all block transformation categories in.

illusaen committed:
Removing extraneous transforms.native.js file. Adding a fix for transforming group -> column block that resulted in crash.

illusaen commented on packages/block-editor/src/components/block-list/index.native.js:
```diff
@@ -60,7 +60,7 @@ export class BlockList extends Component {
 			renderAppender: this.props.renderAppender,
 			onDeleteBlock: this.props.onDeleteBlock,
-			contentStyle: this.props.contentstyle,
+			contentStyle: this.props.contentStyle,
```
Note: this is an unrelated (bug?) I found that doesn't seem to change anything when changed. However, `contentstyle` (original with no caps) ONLY appears in this one case, while `contentStyle` appears everywhere else (and is the actual prop for the RN `View`).

illusaen commented on packages/block-library/src/column/edit.native.js:
```diff
@@ -52,6 +52,10 @@ function ColumnEdit( {
 	clientId,
 	blockWidth,
 } ) {
+	if ( ! contentStyle ) {
```
NOTE: When transforming a group -> column, `contentStyle` is undefined, which leads to a crash. I added this little kludge and the transformation magically works as the web does, meaning that the width in `contentStyle` doesn't actually matter/is reset?

I'm 100% open to ideas as to WHY `contentStyle` is undefined when passing through `createBlock` even though my debugging shows that creating a new column (which somehow gives a `contentStyle` to `column/edit.native.js`) doesn't set `contentStyle` anywhere in its `props`/`attributes`/etc either.

However, I'm not sure that this is a bug that should be something fixed in this PR.

iamthomasbishop commented:
@illusaen This is looking pretty solid and should set us up nicely for the next iteration of the UI, nice work! 

I tested out the latest test build on iOS and broke my feedback down into 2 sections (plus one indirectly-related “bonus” bug that I discovered while testing): things that are mobile-specific and things that aren’t really “bugs” because behavior matches that of the web UI.

## Mobile-specific things

- On the initial block options sheet (shown after user taps •••), can we append ellipses to the ”Transform block” action label? eg. “Transform block…”?
- I don’t think this is possible, at least on iOS with the current ActionSheet component, but I wanted to ask just to be sure: Would it be possible to add block icons in the “Transform X to” Action/BottomSheet? ([Example 1](https://cloudup.com/cqy0mdTBew6), [Example 2](https://cloudup.com/cf2qmq7lnYf))

## Not mobile specific

These are things that also apply to the web editor, so I wouldn’t necessarily consider them ”bugs”:

- It’d be great if we could retain some of the text-based blocks’ alignment properties, but I’m not sure how realistic this is. For example, when I tried to transform a center-aligned Paragraph block to a Heading, I expected the alignment setting to be retained.
- Should we allow More and Page Break to be transformed between one another?
- The behavior of transforming a Pullquote into a Heading feels a bit off. It essentially just pulls the quote part into a separate/new Heading block (which is added above the Pullquote block) and the citation part stays in place. Again, same behavior on web, so not a “bug“ per se ([Example](https://cloudup.com/c_GBVMdmzmL))
- I found it a bit odd that when transforming a media block like Image or Video to File that the title of the file wasn’t applied in the “File name” input, but this is also the case on web. The one diff on the web is that when you do this, the “File name” input is auto-focused. Any chance we can also focus that input upon successful transform? Relatedly, would it be possible to attach the same auto-focus behavior on the caption input after a Cover block is transformed to an Image, Video, etc? 
- I found it odd that (even on web) Video and Audio blocks weren’t options when transforming from a File block, but I don’t necessarily see this as something we should fork off on mobile to support
- In some cases (List, Quote, Pullquote, etc.) transforming a block results in breaking the contents into separate blocks. Would it make sense to append “blocks” to the labels on the ActionSheet? i.e. Transform List to: “Paragraph blocks” and “Heading blocks” in these cases?

## Unrelated

- I also found one unrelated issue where Cover blocks with a video background seem to be throwing an error in the UI (“problem displaying block” on the block). Regular ol’ Video blocks seem to be fine.
- Did we lose a separator on the Heading block’s block toolbar? ([Example](https://cloudup.com/cIVDCEPSD4k))

chipsnyder reviewed (approved):
Very nice work on this @illusaen 👍 

I like how you handled the `supportedMobileTransforms` and grouped them in the `transformationCategories`.

The code looks good and I didn't have anything to add in addition to @iamthomasbishop's design review.

illusaen commented:
@iamthomasbishop Thanks! :)

>     * On the initial block options sheet (shown after user taps •••), can we append ellipses to the ”Transform block” action label? eg. “Transform block…”?
Done, see new GIF attached below :)

>     * I don’t think this is possible, at least on iOS with the current ActionSheet component, but I wanted to ask just to be sure: Would it be possible to add block icons in the “Transform X to” Action/BottomSheet? ([Example 1](https://cloudup.com/cqy0mdTBew6), [Example 2](https://cloudup.com/cf2qmq7lnYf))
Unfortunately not possible to add images to `ActionSheetIOS` on RN.

> ## Not mobile specific
>     * It’d be great if we could retain some of the text-based blocks’ alignment properties, but I’m not sure how realistic this is. For example, when I tried to transform a center-aligned Paragraph block to a Heading, I expected the alignment setting to be retained.
>     * Should we allow More and Page Break to be transformed between one another?
>     * The behavior of transforming a Pullquote into a Heading feels a bit off. It essentially just pulls the quote part into a separate/new Heading block (which is added above the Pullquote block) and the citation part stays in place. Again, same behavior on web, so not a “bug“ per se ([Example](https://cloudup.com/c_GBVMdmzmL))
>     * I found it odd that (even on web) Video and Audio blocks weren’t options when transforming from a File block, but I don’t necessarily see this as something we should fork off on mobile to support
So for the items above, I'm using directly web's transformation functions. I could add the functionality but it would require kludging it, so I'm not sure if we should discuss just changing the behaviour for both platforms on a separate thread?

Edit: I 100% agree with these ideas on UX btw. I also found it weird that transforming Group -> Column results in that group **wrapped** in a column instead of turning each item in the group into a column.

>     * I found it a bit odd that when transforming a media block like Image or Video to File that the title of the file wasn’t applied in the “File name” input, but this is also the case on web. The one diff on the web is that when you do this, the “File name” input is auto-focused. Any chance we can also focus that input upon successful transform? Relatedly, would it be possible to attach the same auto-focus behavior on the caption input after a Cover block is transformed to an Image, Video, etc?
Looking into this now!

>     * In some cases (List, Quote, Pullquote, etc.) transforming a block results in breaking the contents into separate blocks. Would it make sense to append “blocks” to the labels on the ActionSheet? i.e. Transform List to: “Paragraph blocks” and “Heading blocks” in these cases?
Framework for this added; just need list of items to add to it :) In new GIF (attached below) should be able to see the one I added for testing purposes, List > Paragraph.

> ## Unrelated
> 
>     * I also found one unrelated issue where Cover blocks with a video background seem to be throwing an error in the UI (“problem displaying block” on the block). Regular ol’ Video blocks seem to be fine.
Added as bug.

>     * Did we lose a separator on the Heading block’s block toolbar? ([Example](https://cloudup.com/cIVDCEPSD4k))
Added as bug.

<details>
  <summary>GIF of ellipses + added "blocks" word</summary>
  
![string-change](https://user-images.githubusercontent.com/13263478/113221624-0809e200-924b-11eb-8355-5f778705bbbb.gif)

</details>

illusaen committed:
Adding ' blocks' to end of transformations which split one block into multiple.

illusaen commented:
@iamthomasbishop I think I've gotten the changes in; planning to merge at EOD today. If you see something please feel free to tell me and I'll make a separate PR for it!
